src/query_graph/graph_path.rs

```diff
@@ -1338,12 +1340,97 @@ impl OpGraphPath {
+            .and_then(|n| n.fields.get(&itf.field_name))
+        else {
+            // Should not happen
+            return Err(FederationError::internal(
+                "Unable to find interface field ({itf}) in schema: {schema}",
+            ));
+        };
+        for implem in valid_schema.possible_runtime_types(comp_type_pos)? {
+            let ty = valid_schema
+                .get_type(implem.type_name.clone())?
+                .get(schema)?;
+            if !ty.directives().has("key") {
+                continue;
+            }
+            if !field.directives.has("shareable") {
+                continue;
+            }
+            let base_ty_name = field.ty.inner_named_type();
+            if is_leaf_type(schema, base_ty_name) {
+                continue;
+            }
+            match schema.get_object(&field.name) {
+                Some(ty)
+                    if ty
+                        .fields
+                        .values()
+                        .all(|ty| is_leaf_type(schema, ty.ty.inner_named_type())) =>
+                {
+                    for node in self.graph.nodes_for_type(&ty.name).cloned() {
+                        let node = &self.graph.graph()[node];
+                        let tail = &self.graph.graph()[self.tail];
+                        if node.source == tail.source {
+                            continue;
+                        }
+                        let node_ty = match &node.type_ {
+                            QueryGraphNodeType::SchemaType(ty) => ty,
+                            QueryGraphNodeType::FederatedRootType(_) => {
+                                return Err(FederationError::internal(format!(
+                                    "{implem} is an object in {} but a {} in {}",
+                                    tail.source, node.type_, node.source
+                                )))
+                            }
+                        };
+                        let node_ty_name = node_ty.type_name();
+                        let node_ty = valid_schema.get_type(node_ty_name.clone())?.get(schema)?;
+                        let is_shareable = node_ty.directives().has("shareable");
+                        let fields = match node_ty {
+                            ExtendedType::Object(obj) => obj.fields.keys(),
+                            ExtendedType::Interface(int) => int.fields.keys(),
+                            _ => {
+                                return Err(FederationError::internal(format!(
+                                    "{implem} is an object in {} but a {} in {}",
+                                    tail.source, node.type_, node.source
+                                )))
+                            }
+                        };
+                        if !is_shareable || !fields.clone().any(|f| f == &itf.field_name) {
+                            continue;
+                        }
+                        if node_ty_name != base_ty_name {
```

Comment:
We check that `node_ty` is composite in the match statement above, when we get an iterator over the fields' keys.

Reply:
It's slightly different. take this abridged schema:
```graphql
type OutputType # fields omitted, implementation detail
type Object {
  field: OutputType
}
```
`node_ty` refers to the `Object` type. That's what's checked above. The JS additionally checks if `OutputType` is an object or interface. The comment above the function does explain the background for why this is necessary, though it's definitely hard to wrap your head around. I tried to write an explanation in my own words but it wasn't much clearer so 💀 . Maybe we can go through this in a fed standup.

The thing that isn't explicitly written in the comment but is true because of how GraphQL works, is that the only difference between different object types is their fields. So `field: OutputType` in subgraph A and `field: SomethingElse` in subgraph B is not actually a difference on its own if `OutputType` and `SomethingElse` are both object types or interfaces. It's only different if the available fields in `OutputType` and `SomethingElse` are different. And that, we don't need to check inside this function, because the function would be called separately on the inner types. So while inside this function, detecting that two implementers of an interface have the same field but different *scalar* outputs (eg. one is `String` and one is `MyCustomScalar`), means that the implementers are definitely different. If two implementers have the same field with different *object* outputs, it actually doesn't mean those implementers are really different from each other. I'm repeating myself a bit but I hope that makes some sense (and that I didn't make a mistake ✅ ).